### PR TITLE
Editor: Layer renaming

### LINKED ISSUE #9248: Nameable Layers
## Summary

Discussed and requested by design in conjunction with #8983 but isolated to make sure details are attended to accordingly.

Needs some discovery and syncing cross pods before this is dev ready. My understanding is this would be adding functionality that adds a new property getting saved to every story layer that would let them have a readable name separate from their id, etc. 

## References

https://www.figma.com/file/dc9xrVlRhQq4KTE9WQzU1R/_Editor-Sprint-Feature-Level-Designs_?node-id=9935%3A421370
Screen Shot 2021-10-04 at 1 10 00 PM

## Acceptance Criteria

- [ ] You can rename the layer by double-clicking on it in the Layers Panel, using Canvas RCM->Rename layer or Layers Panel context menu.
- [ ] In edit mode: Enter key and blur event will save changes, Escape key will cancel them.

## Other notes 

Naming layers is going to impact keyboard users because right now the layers panel is hidden for keyboards as all functionality is mirrored in keyboard shortcuts. The accessibility of the layers panel will need to be reconsidered ahead of this functionality getting added to the layers panel.

### PR DESCRIPTION
**Pull request by marianakyrkosh**
## Context

Editing layers names

## Summary

Editing of the layers names can be enabled by:
- double clicking on the layers names;
- clicking Rename Layer in the RCM of the element;
- clicking Rename Layer in the RCM of the layers menu;

A name is saved by clicking Enter, clicking outside of the layer button (Blur).
The editing is canceled without saving a new name buy clicking Escape.

## Relevant Technical Choices

layerName property was added to the element object.

## To-do

1. Several components can be extracted from the layers.js file to separate files.
2. Keyboard events handler functions can be extracted to utils so they can be reused.

## User-facing changes

image
image
image

## Testing Instructions

- [ ] This is a non-user-facing change and requires no QA

This PR can be tested by following these steps:

Renaming successful scenario:
1. Double click on the Name -> an input appears -> type a new name -> click Enter -> the layer has a new name.
2. Right click on the element -> a menu opens -> click Rename Layer -> an input appears -> type a new name -> click Enter -> the layer has a new name.
3. Right click on a layer button in the Layers menu -> a menu opens -> click Rename Layer -> an input appears -> type a new name -> click Enter -> the layer has a new name.
4. Repeat 1 - 4, but instead of clicking Enter, click outside of the layer button (Blur).

Renaming canceled scenario:
5. Repeat 1 - 4, but click Escape button instead of Enter -> the layer name should not be updated.

Note: when an input field is visible, layer action icons should be hidden.

## Checklist

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/googleforcreators/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This code is covered by automated tests (unit, integration, and/or e2e) to verify it works as intended ([docs](https://github.com/googleforcreators/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #9248

### CODE REVIEW

File: packages/story-editor/src/components/panels/design/layer/layer.js
Comment (barklund) [inline packages/story-editor/src/components/panels/design/layer/layer.js]: This has to use the current "default" name that is returned by `LayerContent`. Currently this returns a component, but in my PR in #11429 it has been changed to a function, `getLayerText`, returning a simple string.

I have split the relevant part of the code into a new separate PR in #11485. Feel free to merge that one as soon as it is approved and then you can merge main in to update this part here.

File: packages/story-editor/src/components/panels/design/layer/layer.js
Comment (barklund) [inline packages/story-editor/src/components/panels/design/layer/layer.js]: I think the label should be `"Layer name"`, as that is what you're editing in the input.

File: packages/story-editor/src/components/panels/design/layer/useLayerSelection.js
Comment (barklund) [inline packages/story-editor/src/components/panels/design/layer/useLayerSelection.js]: The "detail === 2" part here definitely needs a comment (that it's a double click).

Alternatively, we have a `useDoubleClick` that we use elsewhere for these events.

Comment (barklund) [reply packages/story-editor/src/components/panels/design/layer/useLayerSelection.js]: Also, I think that locked layers can still be renamed. I don't see why not at least. Figma allows this, so I think we should just do that too.

File: packages/story-editor/src/components/panels/design/layer/useLayerSelection.js
Comment (barklund) [inline packages/story-editor/src/components/panels/design/layer/useLayerSelection.js]: This property doesn't belong here at all. The elements in the story are persisted to the database and changes add an entry into the history.

Saving this information directly on the element has a number of issues:

- It adds history entries
- It is stored to the database
- It causes needless re-renders
- It allows multiple layers to be renamed at the same time

See this for some issues, where I have multiple layers with an input field:
Screen Shot 2022-05-12 at 21 26 40

I understand that this information needs to be available in both the canvas and the layer panel, so it has to be stored somewhere "high" in the component tree, but it should not be in the story provider directly. Perhaps it could be a new property in the canvas provider, with state`renamableLayer` and actions `setRenamableLayer` and `clearRenamableLayer` and then `null` means no layer is renamable, and some string means that the layer with the ID matching that string is renamable?

File: packages/story-editor/src/components/panels/design/layer/layer.js
Comment (barklund) [inline packages/story-editor/src/components/panels/design/layer/layer.js:32]: This file is definitely getting too long. I also have additional edits in #114290, so if you have some extra time, it would be refactor this somewhat, alternatively just create a P2 follow-up ticket to refactor this component.

File: packages/story-editor/src/app/rightClickMenu/items/layerName.js
Comment (barklund) [inline packages/story-editor/src/app/rightClickMenu/items/layerName.js]: This should again not go through the story provider, but some other provider. Also, why is the editable property set to the opposite of the locked property? And why not just use `!`?

Comment (barklund) [reply packages/story-editor/src/app/rightClickMenu/items/layerName.js]: Oh yeah, and we need the property, wherever it ends up existing, to be called something like `renamable` or `changeLayerName` or something like that. `editable` is too broad a term, that we also use for other things, so it can get confusing (we have an edit-mode for elements, but it is completely unrelated to this property).

Comment (barklund) [reply packages/story-editor/src/app/rightClickMenu/items/layerName.js]: Finally, the input field needs to be focused, and the layer panel needs to be opened if not already opened, when you right click to rename an element.

This should be done after the refactoring to a property on (e.g.) the canvas provider and then has to be done in two places:

1. In the footer layer button in `components/footer/layers`, the `isOpen` property should be set to true in an effect, if `renamableLayer` is non-null. That would force-open the dialog.
2. In the layer in `components/panels/design/layer/layer.js`, the input field should be focused in an effect, if `renamableLayer` is set to the specific layer. That would force-focus the input and it would work even if the layer panel was originally hidden.

File: packages/story-editor/src/components/panels/design/layer/layer.js
Comment (barklund) [reply packages/story-editor/src/components/panels/design/layer/layer.js:32]: Come to think of it, the whole layer panel should even live inside `panels/design/` at all anymore. It's hardly a panel (doesn't really use any panel bits, except for an unpadded `PanelContent`, which could be removed), and it's definitely not related to the design panel at all.

So, follow-up ticket to refactor this component as well as move the whole layer dialog to a fresh location (probably as `components/layers` or something like that).

File: packages/story-editor/src/components/panels/design/layer/layer.js
Comment (barklund) [reply packages/story-editor/src/components/panels/design/layer/layer.js]: Actually, the layer name, if provided, should be returned by all those `getLayerText` functions in all the different elements. It's not very DRY to go around adding this to 5 different functions.

Actually a better way to do it would be to create a single `getLayerName` function in the elements package, that would first look at `element.layerName` and if not set, use the layer name for the element type. This could also factor in the background layer name too. That's a much cleaner way to do it actually! I will amend #11485 to do that instead!

That would result in the custom layer name as entered by the user would also be used in the right-click select menu and as the aria-label for the element on the canvas (neither of which works currently).

See below, where the layer name is not used in the right-click menu:
Screen Shot 2022-05-12 at 21 43 05

Comment (barklund) [reply packages/story-editor/src/components/panels/design/layer/layer.js]: I did this slight refactor in 57c7b283bc.

**Review by barklund — DISMISSED** [review]
This is really good, but just has a few issues.

- The current layer name should be extracted from the element type, but this work is already done in #11485, so feel free to merge that in once merged to main (@merapi will probably do that first thing Friday)
- The state of whether a layer is renamable should not exist in the story provider, but in some other location. Also, only one layer should ever be renamable.
- If a layer is renamed, the input should be focused (and the layer dialog should be open to do that). It should never be possible to have the rename input unfocused at all.

I have suggested fixes for all these inline below. I will tentatively approve pending the implementation of the above changes.

File: packages/story-editor/src/components/panels/design/layer/layer.js
Comment (marianakyrkosh) [reply packages/story-editor/src/components/panels/design/layer/layer.js]: It's 'Renaming a layer.' in the Figma  design (the link to the design is in the ticket).

File: packages/story-editor/src/components/panels/design/layer/useLayerSelection.js
Comment (merapi) [reply packages/story-editor/src/components/panels/design/layer/useLayerSelection.js]: +1 for `renamableLayer` (id of the layer currently being renamed). CanvasProvider can do the job (spans canvas and layers panel).

File: packages/story-editor/src/components/panels/design/layer/layer.js
Comment (marianakyrkosh) [reply packages/story-editor/src/components/panels/design/layer/layer.js]: Done

File: packages/story-editor/src/components/panels/design/layer/useLayerSelection.js
Comment (marianakyrkosh) [reply packages/story-editor/src/components/panels/design/layer/useLayerSelection.js]: Done

File: packages/story-editor/src/components/panels/design/layer/layer.js
Comment (merapi) [reply packages/story-editor/src/components/panels/design/layer/layer.js]: `Renaming a layer.` is just an example layer name in Figma, right? the input itself holds the layer name, so `Layer name` seems to be a good label for it.

File: packages/story-editor/src/components/panels/design/layer/useLayerSelection.js
Comment (merapi) [reply packages/story-editor/src/components/panels/design/layer/useLayerSelection.js]: @barklund WDYT about renaming the background layer? (`!isBackground` is now removed). Should we ask UX?

File: packages/story-editor/src/components/panels/design/layer/useLayerSelection.js
Comment (marianakyrkosh) [reply packages/story-editor/src/components/panels/design/layer/useLayerSelection.js]: Moved code to the canvasProvider and fixed multiple renamable layers bug.

File: packages/story-editor/src/app/rightClickMenu/items/layerName.js
Comment (marianakyrkosh) [reply packages/story-editor/src/app/rightClickMenu/items/layerName.js]: Done

**Review by merapi — CHANGES_REQUESTED** [review]
I'm reporting what I see in the current version:

1. After entering the edit mode (via double click) you cannot select a part of the text, instead it cancels the edit mode.

2. Go into edit mode, type something and cancel the edit mode (press Escape). The layer name is reverted to the old name - good, but when you enter the edit mode again you see the old (canceled) value.

3. image
This^ should look like it looked before (and in Figma). Those extra icons are unnecessary (I deleted my layer by accident when selecting text).

4. If the Background layer name cannot be changed then you shouldn't be able to enter the edit mode on that layer.

File: packages/story-editor/src/components/panels/design/layer/layer.js
Comment (barklund) [reply packages/story-editor/src/components/panels/design/layer/layer.js:32]: * Added new issue #11576

**Review by merapi — DISMISSED** [review]
This looks good now 👍🏻 

Nit: handling empty layer name.
Right now if you add layer `media/unsplash:gl8BgGhvVdg`, rename it to `layer1`, then rename it to empty string, it will become `media/unsplash:gl8BgGhvVdg`, but it should probably become `layer1`? Then if you enter the edit mode again - it will show an empty input (should always be the layer name visible before entering the edit mode?).

And a thing to consider: if we remove the `margin-left: 2px`, then the layer name will not shift when you enter the edit mode - it looks better IMO (even if it's 2px more over the icon).

**Review by merapi — APPROVED** [review]

**Pull request merged**
